browser-polyfill.js

```diff
@@ -0,0 +1,340 @@
+     * therefore be assumed to behave as a Promise.
+     *
+     * @param {*} value The value to test.
+     * @returns {boolean} True if the value is thenable.
+     */
+    const isThenable = value => {
+      return value && typeof value === "object" && typeof value.then === "function";
+    };
+
+    /**
+     * Creates and returns a function which, when called, will resolve or reject
+     * the given promise based on how it is called:
+     *
+     * - If, when called, `chrome.runtime.lastError` contains a non-null object,
+     *   the promise is rejected with that value.
+     * - If the function is called with exactly one argument, the promise is
+     *   resolved to that value.
+     * - Otherwise, the promise is resolved to an array containing all of the
+     *   function's arguments.
+     *
+     * @param {object} promise
+     *        An object containing the resolution and rejection functions of a
+     *        promise.
+     * @param {function} promise.resolve
+     *        The promise's resolution function.
+     * @param {function} promise.rejection
+     *        The promise's rejection function.
+     *
+     * @returns {function}
+     *        The generated callback function.
+     */
+    const makeCallback = promise => {
+      return (...callbackArgs) => {
+        if (chrome.runtime.lastError) {
+          promise.reject(chrome.runtime.lastError);
+        } else if (callbackArgs.length === 1) {
+          promise.resolve(callbackArgs[0]);
+        } else {
+          promise.resolve(callbackArgs);
+        }
+      };
+    };
+
+    /**
+     * Creates a wrapper function for a method with the given name and metadata.
+     *
+     * @param {string} name
+     *        The name of the method which is being wrapped.
+     * @param {object} metadata
+     *        Metadata about the method being wrapped.
+     * @param {integer} metadata.minArgs
+     *        The minimum number of arguments which must be passed to the
+     *        function. If called with fewer than this number of arguments, the
+     *        wrapper will raise an exception.
+     * @param {integer} metadata.maxArgs
+     *        The maximum number of arguments which may be passed to the
+     *        function. If called with more than this number of arguments, the
+     *        wrapper will raise an exception.
+     *
+     * @returns {function(object, ...*)}
+     *       The generated wrapper function.
+     */
+    const wrapAsyncFunction = (name, metadata) => {
+      return function asyncFunctionWrapper(target, ...args) {
+        if (args.length < metadata.minArgs) {
+          throw new Error(`Expected at least ${metadata.minArgs} arguments for ${name}(), got ${args.length}`);
+        }
+
+        if (args.length > metadata.maxArgs) {
+          throw new Error(`Expected at most ${metadata.maxArgs} arguments for ${name}(), got ${args.length}`);
+        }
```

Comment:
In the interest of sensible error messages. We change the number of arguments passed to the underlying function, and the error messages don't make sense when they're complaining about the type of an argument that the user didn't explicitly pass.